docs/Config.md
# Configuration
Configuration is done by the Alyssa.cfg file on working directory.

## Configuration file info
- Config file consists of `key value` lines.
- There's no symbols like = or ". You just directly write the key and value. First word till space is key, rest is value.
- Keys are case insensitive.
- If you set a key more than once, first one will be used.
- Lines with # at beginning are comment lines, and will be ignored by server.
- Config is only read at launch of server, you need to restart the server when you do some modification on config.

## List of possible keys and their default values.
### Core
- **htrootpath ./htroot:** Where your htroot is located.
- **respath ./res** Where the resources folder is.
- **Port 80** Ports to listen to. Multiple ports can be listened by doing i.e. `Port 80 800 8000...`
- **htrespath /res** Where the resources located as HTTP relative path. Unrelated to `respath`. 
Changing is not necessary unless your dynamic content stack conflicts with it.
- **ErrorPages 1** 0 disables error pages completely, 1 enables the server-synthetized error pages, 2 enables the custom error pages that's put on resources directory.
- **Logging 0** Enables/disables logging on Alyssa.log file on working directory.
- **PrintConnections 0** Enables/disables printing connections on console.
- **ColorOutput 1** Enables/disables colorful output on console.
- **IPv6 0** Enables/disables IPv6 support.
- **VirtualHosts [no value]** Location of [Virtual Hosts Configuration](VHost.md) file.
### Additional features
- **DirectoryIndex 1** Enables/disables [directory index pages](res/DirIndex.png).
- **CustomActions 0** 0 disables custom actions completely, 1 enables it, 2 enables it recursively
### SSL-related
- **EnableSSL 0** Enables/disables SSL support.
- **SSLCertPath ./crt.pem** Path of SSL certificate
- **SSLKeyPath ./key.key** Path of SSL key
- **SSLPort 443** Same as Port above but for HTTPS.
- **HTTP2 0** Enables/disables HTTP/2 support. Depends on SSL and server compiled with HTTP/2 support.
- **HSTS 0** Enables/disables HTTP Strict Transport Security.
### CORS/CSP
- **CORSAllowOrigin [no value]** Sets the value of `Access-Control-Allow-Origin` header.
- **CSPHeaders [no value]** Sets the value of `Content-Security-Policy` header.
### Ratelimiting
- **RatelimitInterval 0:** Ratelimit interval. If more amount than specified threshold of request will be made, 
ratelimit will be applied for specified milliseconds. Default is 1000 when unspecified.
- **RatelimitThreshold 0:** Threshold that I said above, default is 10 when unspecified.
- **RatelimitAmount 0:** Amount of milliseconds to defer. default is 150 when unspecified.
- **PollPeriod 0:** Period of poll() calls. (will defer each call that much of milliseconds, 
think that like the ratelimit but for new connections.)


src/Config.cpp
#ifndef AlyssaHeader
#include "Alyssa.h"
#endif

std::vector<string> configcache, configcache_value; char delimiter; bool isCRLF = 0; 

bool Config::Configcache() {//This function reads the config file and caches all the keys and values on the file to 2 separate string arrays. Much easier and faster than reading the same file again and again.
	std::ifstream conf; conf.open("Alyssa.cfg"); conf.imbue(std::locale(std::locale(), new std::codecvt_utf8<wchar_t>));
	if (!conf) { 
		return 0; 
	}
	string buf(std::filesystem::file_size("Alyssa.cfg"), '\0');
	conf.read(&buf[0], std::filesystem::file_size("Alyssa.cfg")); buf += "\1"; string temp = "";
	for (size_t i = 0; i < buf.size(); i++) {
		if (buf[i] < 32) {
			if (temp[0] == '#' || temp[0] == '/') {}//Comment line, discard
			else if (temp == "") {}//Blank line, discard
			else {
				for (size_t i = 0; i < temp.size(); i++) {
					if (temp[i]==' ') {
						configcache.emplace_back(ToLower(Substring(&temp[0], i))); 
						configcache_value.emplace_back(Substring(&temp[0], 0, i + 1));
					}
				}
			}
			temp = ""; if (buf[i + 1] < 32) i++;//CRLF
		}
		else temp += buf[i];
	}
	conf.close(); return 1;
}

string Config::getValue(std::string key, std::string value) {//Interface function for getting a value from config. If value isn't found then value variable on this function will be returned as default value.
	for (size_t i = 0; i < configcache.size(); i++) {
		if (configcache[i] == key) {
			return configcache_value[i];
		}
	}
	return value;
}

bool Config::initialRead() {//Initial read of the config file and setup of setting variables at the startup of the program.
	if (!Configcache()) return 0;
	portStr = getValue("port", "80")+'\1';
	string temp = ""; port.clear();
	for (size_t i = 0; i < portStr.size(); i++) {
		if (portStr[i] >= 48) temp += portStr[i];
		else {
			try {
				port.emplace_back(stoi(temp));
			}
			catch (std::invalid_argument&) {
				ConsoleMsg(0, "Config: ", "invalid port specified on config."); exit(-3);
			}
			if (port[port.size() - 1] > 65535) {
				ConsoleMsg(0, "Config: ", "invalid port specified on config."); exit(-3);
			}
			temp.clear();
		}
	}
	htroot = getValue("htrootpath", "./htroot");
	respath = getValue("respath", "./res");
	htrespath = getValue("htrespath", "/res");
	_htrespath = '.' + htrespath;
	foldermode = stoi(getValue("directoryindex", "0"));
	errorpages = stoi(getValue("errorpages", "0"));
	//whitelist = getValue("whitelist", "");
	logOnScreen = stoi(getValue("printconnections", "0"));
	std::string defaultCorsAllowOrigin = getValue("corsalloworigin", "");
	if (defaultCorsAllowOrigin != "") {
		corsEnabled = 1; defaultCorsAllowOrigin += " ";
		short off = 0, pos = 0;
		while ((off = defaultCorsAllowOrigin.find(' ', off + 1)) >= 0) {
			ACAOList.emplace_back(defaultCorsAllowOrigin.substr(pos, off - pos)); pos = off;
		}
	}
	CSPHeaders = getValue("cspheaders", "");
	if (CSPHeaders != "") CSPEnabled = 1;
	logging = stoi(getValue("logging", "0"));
	EnableIPv6 = stoi(getValue("ipv6", "0"));
	ColorOut = stoi(getValue("coloroutput", "1"));
	VHostFilePath = getValue("virtualhosts", "");
	if (VHostFilePath != "") HasVHost = 1;
	pollPeriod = stoi(getValue("pollperiod", "0"));
	ratelimit_ts = stoi(getValue("ratelimitthreshold", "0"));
	ratelimit_int = stoi(getValue("ratelimitinterval", "0"));
	ratelimit_ms = stoi(getValue("ratelimitamount", "0"));
	if (ratelimit_ms || ratelimit_int || ratelimit_ts) {// If any of ratelimit settings are enabled, check if others has a value too and if they not set them to default.
		if (!ratelimit_int) ratelimit_int = 1000;
		if (!ratelimit_ms)  ratelimit_ms = 150;
		if (!ratelimit_ts)  ratelimit_ts = 10;
		ratelimitEnabled = 1;
	}

#ifdef Compile_WolfSSL
	enableSSL = stoi(getValue("enablessl", "0"));
	if (enableSSL) {
		SSLcertpath = getValue("sslcert", "./crt.pem");
		SSLkeypath = getValue("sslkey", "./key.key");
		SSLportStr = getValue("sslport", "443") + '\0';
	}
	temp.clear();
	for (size_t i = 0; i < SSLportStr.size(); i++) {
		if (SSLportStr[i] >= 48) temp += SSLportStr[i];
		else {
			try {
				SSLport.emplace_back(stoi(temp));
			}
			catch (std::invalid_argument&) {
				ConsoleMsg(0, "Config: ", "invalid port specified on config."); exit(-3);
			}
			if (SSLport[SSLport.size() - 1] > 65535) {
				ConsoleMsg(0, "Config: ", "invalid port specified on config."); exit(-3);
			}
			temp.clear();
		}
	}
	HSTS = stoi(getValue("hsts", "0"));
	if (HSTS && !enableSSL) { ConsoleMsg(0, "Config: ", "HSTS is set on config but SSL is not enabled."); HSTS = 0; }
#ifdef Compile_H2
	EnableH2 = stoi(getValue("http2", "0"));
#endif // Compile_H2

#endif //Compile_WolfSSL

#ifdef Compile_CustomActions
	switch (stoi(getValue("customactions", "0"))) {
		case 1:  CAEnabled = 1; break;
		case 2:  CAEnabled = 1; CARecursive = 1; break;
		default: break;
	}
#endif // Compile_CustomActions

#ifdef Compile_locales
	if (getValue("locale", "") == "tr") Locale = 1; else Locale = 0;
#else
	Locale = 0;
#endif //Compile_locales

#ifdef Compile_zlib
	deflateEnabled = stoi(getValue("deflate", "0"));
#endif // Compile_zlib
	configcache.clear(); configcache_value.clear();
	return 1;
}
